tests/test_cases/test_vpi_structure/vpi_structure.py

```diff
@@ -0,0 +1,129 @@
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.captured_output.seek(0)
+        # check if exists
+        if self.update:
+            with open(self.expected_output_file, "w") as f:
+                f.write(self.captured_output.getvalue())
+            return
+
+        if not os.path.exists(self.expected_output_file):
+            raise FileNotFoundError(
+                f"Expected output file {self.expected_output_file} does not exist"
+            )
+
+        with open(self.expected_output_file) as f:
+            expected_output = f.read()
+
+        if self.captured_output.getvalue() != expected_output:
+            raise ValueError(
+                f"Output does not match expected:\n{self.captured_output.getvalue()}"
+            )
+
+
+@cocotb.test()
+async def test_structure(dut):
+    """
+    Tests that name, fullname, handle type, and length match across simulators.
+
+    Not all of them match at the moment, but we can use this to measure progress.
+    """
+    sim_name = cocotb.SIM_NAME.lower().split()[0]
+
+    try:
+        with verify_output(
+            f"test_structure.{sim_name}.out", update=cocotb.plusargs.get("update")
+        ) as f:
+            for obj, depth in iter_module(dut):
+                objlen = (
+                    f"[{get_len(obj)}]"
+                    if isinstance(obj, cocotb.handle.ValueObjectBase)
+                    else ""
+                )
+                objtype = type(obj).__qualname__
+                treeinfo = f"{'  ' * depth}{obj._name}: {objtype}{objlen}"
+                f.print(f"{treeinfo:50} {obj._path}")
+    except FileNotFoundError:
+        cocotb.log.warning(
+            f"No expected output file found for {sim_name}, assuming this is because it's a proprietary simulator. Use +update to update output file"
+        )
+
+
+@cocotb.test(
+    skip=cocotb.SIM_NAME.lower().startswith("riviera"),
+    expect_error=AttributeError if cocotb.SIM_NAME.lower().startswith("icarus") else (),
+)
+async def test_name_matches_iter(dut):
+    """
+    Test name accessibility and handle lengths.
+
+    All of the names accessible through iteration are also accessible through the name.
+    Multiple instances in Python don't corrupt C++ handle lengths, particularly pseudo objects.
+    """
+
+    t = cocotb.handle.HierarchyObject(dut._handle, dut._path)
+
+    # We need to ensure that these are different objects, so the iteration one (dut)
+    # doesn't fill in the _sub_handles for the name one (t)
+    assert id(t) != id(dut)
+
+    objs = [obj for obj, _depth in iter_module(dut)]
+
+    # this gives "RuntimeError: dictionary changed size during iteration" on verilator / intf_arr
+    # for obj, _depth in iter_module(dut):
+
+    for obj in objs:
+        cocotb.log.info(obj._path)
+        objlen = get_len(obj)
+
+        direct_obj = eval(obj._path)
+        assert obj._handle == direct_obj._handle
```

Comment:
I don't think we want to enforce `_handle` being the same at the test level or external to GPI.
The current implementation of GPI with `SINGLETON_HANDLES` will re-use existing handles, but that shouldn't be expected.

Reply:
We have this though
```
def __eq__(self, other: Any) -> bool:
        if not isinstance(other, SimHandleBase):
            return NotImplemented
        return self._handle == other._handle
```